docs/blog/2019-01-04-using-react-context-api-with-gatsby/index.md

```diff
@@ -0,0 +1,273 @@
+
+>
+> One of the few types of alternative theme that adds real value to users is a low light intensity “night mode” theme. Not only is it easier on the eyes when reading in the dark, but it also reduces the likelihood of migraine and the irritation of other light sensitivity disorders. As a migraine sufferer, I’m interested!
+>
+> [Heydon Pickering](https://inclusive-components.design/a-theme-switcher/)
+
+While discussing the different ways to implement this, I came up with React’s new Context API as a possible implementation strategy. Having worked with Context API before, I set to code. However, I soon realized you will not see a ‘root’ in a Gatsby app like an app built with Create React App.
+
+After Googling a bit, I came across the Gatsby Browser APIs. Specifically, the [wrapRootElement](https://www.gatsbyjs.org/docs/browser-apis/#wrapRootElement) API was what I was looking for. This essentially allows you to wrap your root element with a Provider from Redux or Context for instance. Using this, I managed to achieve dark mode for Laccadive IO.
+
+The following is a detailed instruction on how you too can achieve a similar result with Gatsby and React Context!
+
+## Creating the Context file in a new Gatsby project
+
+First of all, you have to initialize a Gatsby project and start it in develop mode.
+
+1.   npm install –global gatsby-cli
+2.  gatsby new gatsby-dark-mode
+3.  cd gatsby-dark-mode
+4.  gatsby develop
+
+Then, create a `context` folder within src and the `ThemeContext.js` file within it.
+
+```js:title=src/contex/ThemeContext.js
+    import React from 'react'
+
+    const defaultState = {
+      dark: false,
+      toggleDark: () => {},
+    }
+
+    const ThemeContext = React.createContext(defaultState)
+
+    export default ThemeContext
+```
+
+`React.createContext` is a new function in React 16.3 and allows you to create a Context object. It accepts a default state, the value which will be used when a component does not have a matching Provider above it in the tree.  The function returns an object with Provider and Consumer properties which we will be using later.
+
+## Modifying the Gatsby Browser file
+
+Next, write the following code within the `gatsby-browser.js` file, which is in the root folder in a Gatsby project:
+```jsx:title=gatsby-browser.js
+    import React from 'react'
+
+    import ThemeContext from './src/context/ThemeContext'
+
+    class ThemeProvider extends React.Component {
+      state = {
+        dark: false,
+      }
+
+      toggleDark = () => {
+        let dark = !this.state.dark
+        localStorage.setItem('dark', JSON.stringify(dark))
+        this.setState({ dark })
+      }
+
+      componentDidMount() {
+        this.setState({ dark: JSON.parse(localStorage.getItem('dark')) })
```

Comment:
One cool addition you could do here is show how to adapt to a dark mode setting from the OS, e.g. [like I do here](https://github.com/DSchau/gatsby-mail/blob/master/src/components/theme.js#L10)